### PR TITLE
impr/clients: Handle TLS config and MTLS for logcli and promtail

### PR DESCRIPTION
**Pull request by CyrilPeponnet**
This will add support for TLS and MTLS for both promtail and logcli.

If you want to try it I suggest to use the amazing [mtlsproxy](https://github.com/aporeto-inc/mtlsproxy) and the wonderfull [tls gen tools](https://github.com/aporeto-inc/tg) as follow:

```console
# Generate a CA
tg cert \
    --name "ca-loki" \
    --is-ca \
    --pass "passphrase" \
    --common-name "Loki CA"

# Generate server cert
tg cert \
    --auth-server \
    --org "loki" \
    --name "loki" \
    --dns "localhost" \
    --signing-cert "ca-loki-cert.pem" \
    --signing-cert-key "ca-loki-key.pem" \
    --signing-cert-key-pass "password"

# Generate a client cert
tg cert \
    --auth-client \
    --org "loki" \
    --name "client" \
    --signing-cert "ca-loki-cert.pem" \
    --signing-cert-key "ca-loki-key.pem" \
    --signing-cert-key-pass "password"

```

Then you can run the proxy as:

```console
/mtlsproxy --backend localhost:3100 --cert loki-cert.pem  --cert-key loki-key.pem --listen :8080 --clients-ca ca-loki-cert.pem --mode tcp
```

And now you can run either `logcli` or `promtail` as:

Logcli:

```console
logcli --addr=https://localhost:8080 --ca-cert=ca-loki-cert.pem --cert client-cert.pem --key client-key.pem query '{app="foo"}'
```

Promtail ~~either via configuration or args~~ through configuration only:

```yaml
server:
  http_listen_port: 9080
  grpc_listen_port: 0

positions:
  filename: /tmp/positions.yaml

clients:
  - url: http://localhost:3100/api/prom/push

    # Sets the `Authorization` header on every promtail request with the
    # configured username and password.
    # password and password_file are mutually exclusive.
    basic_auth:
      username: <string>
      password: <secret>
      password_file: <string>

    # Sets the `Authorization` header on every promtail request with
    # the configured bearer token. It is mutually exclusive with `bearer_token_file`.
    bearer_token: <secret>

    # Sets the `Authorization` header on every promtail request with the bearer token
    # read from the configured file. It is mutually exclusive with `bearer_token`.
    bearer_token_file: /path/to/bearer/token/file

    # Configures the promtail request's TLS settings.
    tls_config:
      # CA certificate to validate API server certificate with.
      # If not provided Trusted CA from sytem will be used.
      ca_file: <filename>

      # Certificate and key files for client cert authentication to the server.
      cert_file: <filename>
      key_file: <filename>

      # ServerName extension to indicate the name of the server.
      # https://tools.ietf.org/html/rfc4366#section-3.1
      server_name: <string>

      # Disable validation of the server certificate.
      insecure_skip_verify: <boolean>

    # Optional proxy URL.
    proxy_url: <string>

scrape_configs:
- job_name: system
  entry_parser: raw
  static_configs:
  - targets:
      - localhost
    labels:
      job: varlogs
      __path__: /var/log/*log
```

Tested:

- [X] no tls both promtail / logcli with query and tail (websocket)
- [X] tls both promtail / logcli with query and tail (websocket)
- [X] mtls both promtail / logcli with query and tail (websocket)

### CODE REVIEW

**Comment by gouthamve** [comment]
Nice! This looks good, but can you use this:https://github.com/prometheus/common/blob/master/config/http_config.go#L123-L131 to configure and create the client?

This has more options (BasicAuth, BearerToken) and is generally used across all prometheus eco-system, best to re-use it here too.

**Comment by CyrilPeponnet** [comment]
@gouthamve sure, I didn't see that it was there in the first place. Few things though:

1 - It doesn't handle encrypted private keys
2 - it doesn't use the SystemCertPool so any tls settings MUST provide the CA

For (1) I'm ok to drop the support of encrypted keys. For (2) I can do the changes to use SystemCertPoll.

**Comment by gouthamve** [comment]
@CyrilPeponnet I've talked to the maintainer of the upstream package and it was suggested that the systemcertpool is automatically picked up. Can you verify that? If not, can you open an issue in `prometheus/common`?

Further, instead of us supporting only TLS config, can we just use the YAML from the `HTTPClientConfig` itself, supporting the full range of options there?

**Comment by CyrilPeponnet** [comment]
1/ that is correct. Default ca is system based my bad:

```   // RootCAs defines the set of root certificate authorities
        // that clients use when verifying server certificates.
        // If RootCAs is nil, TLS uses the host's root CA set.
        RootCAs *x509.CertPool
``` 

2/ for Promtail yes doable for logcli there is no config file.

**Comment by CyrilPeponnet** [comment]
@gouthamve done (2). promtail clientS (as now you can have several clients), can use the prometheus http client config and follows the same validation rules.

File: cmd/promtail/main.go [hunk 1]
```diff
@@ -37,6 +37,12 @@ func main() {
 		}
 	}
 
+	err := config.ClientConfig.Client.Validate()
+	if err != nil {
+		level.Error(util.Logger).Log("msg", "error validating client config", "error", err)
+		os.Exit(1)
+	}
```
Comment (tomwilkie) [inline cmd/promtail/main.go]: Could you move this code into promtail/client/client.go:79 (New) please?  That way other users will benefit from it.

File: docs/logcli.md [hunk 2]
```diff
@@ -44,26 +44,31 @@ Common labels: {job="cortex-ops/consul", namespace="cortex-ops"}
 
 ### Configuration
 
-
 Configuration values are considered in the following order (lowest to highest):
+
```
Comment (tomwilkie) [inline docs/logcli.md]: Please avoid superfluous changes like this

Comment (tomwilkie) [reply docs/logcli.md]: Unless this is invalid markdown?

File: docs/operations.md [hunk 3]
```diff
@@ -5,7 +5,7 @@ This page lists operational aspects of running Loki in alphabetical order:
 ## Authentication
 
 Loki does not have an authentication layer.
-You are expected to run an authenticating reverse proxy in front of your services, such as an Nginx with basic auth or an OAuth2 proxy.
+You are expected to run an authenticating reverse proxy in front of your services, such as an Nginx with basic auth or an OAuth2 proxy, mutual TLS. See [here](promtail-setup.md#custom-clients-options) for more details about supported authentication methods.
```
Comment (tomwilkie) [inline docs/operations.md]: I'm not sure `, mutual TLS.` makes grammatical sense.  Probably best omitted.

And in the docs we try and put new sentences on new lines to help with future diffs.

File: docs/promtail-setup.md [hunk 4]
```diff
@@ -166,3 +169,52 @@ scrape_configs:
         replacement: /your_log_file_dir/$1.log
       ...
 ```
+
+### Custom Clients options
+
+`promtail` clients configuration uses the [prometheus http client](https://godoc.org/github.com/prometheus/common/config) implementation.
```
Comment (tomwilkie) [inline docs/promtail-setup.md]: s/prometheus/Prometheus/

File: pkg/promtail/client/client.go [hunk 5]
```diff
@@ -85,9 +87,19 @@ func New(cfg Config, logger log.Logger) Client {
 
 		externalLabels: cfg.ExternalLabels,
 	}
+
+	var err error
+	c.client, err = config.NewClientFromConfig(cfg.Client, "promtail")
+	if err != nil {
+		level.Error(c.logger).Log("msg", "error while creating http client", "error", err) //nolint
```
Comment (tomwilkie) [inline pkg/promtail/client/client.go]: We tend not to log in "leaf" functions - the error will get logger at the top of the call tree, ie in the `main` func.

You should do an `error.Wrap` if you want to add more context.

File: pkg/promtail/client/client.go [hunk 6]
```diff
@@ -137,7 +149,7 @@ func (c *client) run() {
 func (c *client) sendBatch(batch map[model.Fingerprint]*logproto.Stream) {
 	buf, err := encodeBatch(batch)
 	if err != nil {
-		level.Error(c.logger).Log("msg", "error encoding batch", "error", err)
+		level.Error(c.logger).Log("msg", "error encoding batch", "error", err) //nolint
```
Comment (tomwilkie) [inline pkg/promtail/client/client.go]: What lint error are these hiding?

File: pkg/promtail/client/multi.go [hunk 7]
```diff
@@ -17,9 +17,13 @@ func NewMulti(logger log.Logger, cfgs ...Config) (Client, error) {
 	if len(cfgs) == 0 {
 		return nil, errors.New("at least one client config should be provided")
 	}
-	var clients []Client
+	clients := []Client{}
```
Comment (tomwilkie) [inline pkg/promtail/client/multi.go]: This change is superfluous, the previous code id valid.

**Comment by tomwilkie** [comment]
Thanks @CyrilPeponnet, this looks really good!  I'm added a bunch of nit-picky comments but we should be able to get this merge soon.

File: pkg/promtail/client/client.go (hunk 6 above)
Comment (CyrilPeponnet) [reply pkg/promtail/client/client.go]: forget it I may have dreamt.

**Comment by CyrilPeponnet** [comment]
@tomwilkie ready for another round :)

File: docs/promtail-setup.md [hunk 8]
```diff
@@ -166,3 +169,87 @@ scrape_configs:
         replacement: /your_log_file_dir/$1.log
       ...
 ```
+
+### Custom Client options
+
+`promtail` client configuration uses the [Prometheus http client](https://godoc.org/github.com/prometheus/common/config) implementation.
+Therefor you can configure the following authentication parameters in the `client` or `clients` section.
```
Comment (Kuqd) [inline docs/promtail-setup.md]:
```suggestion
Therefore you can configure the following authentication parameters in the `client` or `clients` section.
```

**Pull request merged**